vowpalwabbit/core/src/reductions/eigen_memory_tree.cc

```diff
@@ -0,0 +1,935 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "vw/core/reductions/eigen_memory_tree.h"
+
+#include "vw/common/future_compat.h"
+#include "vw/config/options.h"
+#include "vw/core/array_parameters.h"
+#include "vw/core/example.h"
+#include "vw/core/learner.h"
+#include "vw/core/numeric_casts.h"
+#include "vw/core/rand48.h"
+#include "vw/core/rand_state.h"
+#include "vw/core/setup_base.h"
+#include "vw/core/v_array.h"
+#include "vw/core/vw.h"
+#include "vw/io/logger.h"
+
+#include <algorithm>
+#include <cfloat>
+#include <climits>
+#include <cmath>
+#include <cstdio>
+#include <ctime>
+#include <list>
+#include <memory>
+#include <sstream>
+
+using namespace VW::LEARNER;
+using namespace VW::config;
+
+// TODO: This file has several cout print statements. It looks like
+//       they should be using trace_message, but its difficult to tell
+namespace VW
+{
+namespace reductions
+{
+namespace eigen_memory_tree
+{
+///////////////////////Helper/////////////////////////////
+//////////////////////////////////////////////////////////
+float median(std::vector<float>& array)
+{
+  int size = array.size();
+  auto nth = array.begin() + size / 2; 
+
+  if (size % 2 == 0) {
+    std::nth_element(array.begin(), nth  , array.end());
+    std::nth_element(array.begin(), nth-1, array.end());
```

Comment:
>>The other elements are left without any specific order, except that none of the elements preceding nth are greater than it, and none of the elements following it are less.

Looks like this call is invalidating guaraantees about nth from previous line.
We probably should either:
1) get the *nth value between these calls
2) replace second nth_element call with something like (constant factor seems to be better here)
```
auto n_minus_1th = std::max_element(array.begin(), array.begin() + size / 2 - 1)
```

Reply:
I grabbed the nth between the calls. I wasn't entirely sure grabbing the max was guaranteed to work given nth_element. Feel free to change it.